File: problems/src/linked_list_cycle_ii/step3.ts
// Step3
// 目的: 覚えられないのは、なんか素直じゃないはずなので、そこを探し、ゴールに到達する

// 方法
// 時間を測りながらもう一度解く
// 10分以内に一度もエラーを吐かず正解
// これを3回連続でできたら終わり
// レビューを受ける
// 作れないデータ構造があった場合は別途自作すること

/*
  時間計算量: O(1)
  空間計算量: O(1)
*/


// 初期コード
class ListNode {
  val: number
  next: ListNode | null
  constructor(val?: number, next?: ListNode | null) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
  }
}

export function detectCycle(head: ListNode | null): ListNode | null {
  let slow = head;
  let fast = head;
  while (true) {
    if (!fast || !fast.next) {
      return null;
    } 
    slow = slow?.next!;
    fast = fast.next.next;
    if (fast === slow) {
      break;
    }
  }
  slow = head;
  while (fast) {
    if (slow === fast) {
      return fast;
    } 
    slow = slow?.next!;
    fast = fast.next;
  }
  return null;
};

File: problems/src/remove_duplicates_from_sorted_list/step3.rs
// Step3
// 目的: 覚えられないのは、なんか素直じゃないはずなので、そこを探し、ゴールに到達する

// 方法
// 時間を測りながらもう一度解く
// 10分以内に一度もエラーを吐かず正解
// これを3回連続でできたら終わり
// レビューを受ける
// 作れないデータ構造があった場合は別途自作すること

/*
  時間計算量: O(n)
  空間計算量: O(1)
*/

/*
 * @lc app=leetcode id=83 lang=rust
 *
 * [83] Remove Duplicates from Sorted List
 */
/*
 * @lc app=leetcode id=83 lang=rust
 *
 * [83] Remove Duplicates from Sorted List
 */

// @lc code=start
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

impl Solution {
    pub fn delete_duplicates(mut head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut current_node = head.as_mut();
        while let Some(current) = current_node {
            let mut next_node = current.next.take();
            while let Some(next) = next_node.as_mut() {
                if current.val != next.val {
                    current.next = next_node;
                    break;
                }
                next_node = next.next.take();
            }
            current_node = current.next.as_mut();
        }
        head
    }
}
// @lc code=end
